internal/bufpushaction/bufpushaction.go

```diff
@@ -0,0 +1,265 @@
+	moduleName       string
+	githubClient     github.Client
+	bufRunner        commandRunner
+}
+
+var errNoTrackSupport = errors.New("The installed version of buf does not support setting the track. Please use buf v1.0.0 or newer.")
+
+func (p pusher) push(ctx context.Context) error {
+	// versions of buf prior to --track support emit this message when running `buf push --track foo --help`
+	const unsupportedTrackHelpFragment = "unknown flag: --track"
+
+	// make sure --track is supported
+	if p.track != "main" {
+		_, stderr, err := p.bufRunner.Run(ctx, "push", "--track", p.track, "--help")
+		if err != nil {
+			if strings.Contains(stderr, unsupportedTrackHelpFragment) {
+				return errNoTrackSupport
+			}
+			return err
+		}
+	}
+	tags, err := p.getTags(ctx)
+	if err != nil {
+		return err
+	}
+	for _, tag := range tags {
+		if len(tag) != 40 {
+			continue
+		}
+		if _, err := hex.DecodeString(tag); err != nil {
+			continue
+		}
+		status, err := p.githubClient.CompareCommits(ctx, tag, p.currentGitCommit)
+		if err != nil {
+			if github.IsNotFoundError(err) {
+				continue
+			}
+			return err
+		}
+		switch status {
+		case github.CompareCommitsStatusIdentical:
+			p.notice(fmt.Sprintf("Skipping because the current git commit is already the head of track %s", p.track))
+			return nil
+		case github.CompareCommitsStatusBehind:
+			p.notice(fmt.Sprintf("Skipping because the current git commit is behind the head of track %s", p.track))
+			return nil
+		case github.CompareCommitsStatusDiverged, github.CompareCommitsStatusAhead:
+		default:
+			return fmt.Errorf("unexpected status: %s", status)
+		}
+	}
+	stdout, stderr, err := p.bufRunner.Run(ctx, "push", "--track", p.track, "--tag", p.currentGitCommit, p.input)
+	if err != nil {
+		return errors.New(stderr)
+	}
+	if len(stderr) > 0 {
+		p.notice(stderr)
+	}
+	commit := stdout
+	if commit == "" {
+		trackRef := fmt.Sprintf("%s:%s", p.moduleName, p.track)
+		stdout, stderr, err = p.bufRunner.Run(ctx, "beta", "registry", "commit", "get", trackRef, "--format", "json")
+		if err != nil {
+			return errors.New(stderr)
+		}
+		var commitInfo struct {
+			Commit string `json:"commit"`
+		}
+		if err := json.Unmarshal([]byte(stdout), &commitInfo); err != nil {
+			return errors.New("unable to parse commit info")
+		}
+		commit = commitInfo.Commit
+	}
+	p.setOutput("commit", commit)
+	p.setOutput("commit_url", bsrCommitURL(p.moduleName, commit))
+	return nil
+}
+
+func (p pusher) notice(message string) {
+	fmt.Fprintln(p.stdout, workflowNotice(message))
+}
+
+func workflowNotice(message string) string {
+	return fmt.Sprintf("::notice::%s", message)
+}
+
+func (p pusher) setOutput(name, value string) {
+	fmt.Fprintln(p.stdout, workflowOutput(name, value))
+}
+
+func workflowOutput(name, value string) string {
+	return fmt.Sprintf("::set-output name=%s::%s", name, value)
+}
+
+func bsrCommitURL(moduleName, commit string) string {
+	return fmt.Sprintf("https://%s/tree/%s", moduleName, commit)
+}
```

Comment:
I agree. I had it there because it's also called from tests.  I inlined the one call in production code and moved the function to a test file so tests can still use it.

Reply:
I generally like https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html and prefer expectation in test to just be fully written out, rather than sharing small bits of logic among the implementation and test code.

Reply:
fixed